### ISSUE TITLE
BooleanExpressionComplexity complains about similar and simple expression

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#1525]
Base on discussion at https://github.com/checkstyle/checkstyle/pull/1468, BooleanExpressionComplexity (http://checkstyle.sourceforge.net/config_metrics.html#BooleanExpressionComplexity) Check should not raise violation (with special property) of code like this:
```
            return type == TokenTypes.PACKAGE_DEF
                || type == TokenTypes.CLASS_DEF
                || type == TokenTypes.INTERFACE_DEF
                || type == TokenTypes.ENUM_DEF
                || type == TokenTypes.ANNOTATION_DEF
                || type == TokenTypes.METHOD_DEF
                || type == TokenTypes.CTOR_DEF
                || type == TokenTypes.VARIABLE_DEF
```
It is not always possible to refactor that code and in current form performance is the same be readability is better.

 before implementation, please generate report for big projects to prove that such code is common.

**Comment by MEZk** [comment]
@romani 
I'm on it

**Comment by MEZk** [comment]
@romani

Config 
```xml
<module name="BooleanExpressionComplexity">
    <property name="max" value="7"/>
    <property name="tokens" value="LAND, LOR"/>
</module>
```

Reports https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/

Reports show that the above mentioned code is not a common pattern, but exists in the code base of the really big projects such as OpenJDK8, Spring, Struts.

**Comment by romani** [comment]
@MEZk , you just found cases where number is bigger than 7, some cases are combination of very different sub-expressions, such cases should continue to be a violation.

But yes, from this report we could see real code which Check should address, starting point for us to analyse problem a bit dipper.

**Comment by MEZk** [comment]
> you just found cases where number is bigger than 7, some cases are combination of very different sub-expressions, such cases should continue to be a violation.

The original problem was found in https://github.com/checkstyle/checkstyle/pull/1468 when we set 'max' property of BooleanExpressionComplexity check to 7. So I decided to do the same and analyze the login in return statements. 

These cases show that the above mentioned code pattern can be found in the real projects. The examples are a food for thought.
1) [example 1](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/spring-framework/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/spring-framework/spring-jdbc/src/main/java/org/springframework/jdbc/support/JdbcUtils.java.html#L411)
2) [example 2](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/apache-struts/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/apache-struts/plugins/embeddedjsp/src/main/java/org/apache/struts2/el/lang/ELSupport.java.html#L416)
3) [example 3](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/apache-struts/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/apache-struts/plugins/json/src/main/java/org/apache/struts2/json/JSONPopulator.java.html#L133)
4) [example 4](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/spring-framework/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/spring-framework/spring-beans/src/main/java/org/springframework/beans/BeanUtils.java.html#L520)
5) [example 5](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/spring-framework/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/spring-framework/spring-oxm/src/main/java/org/springframework/oxm/jaxb/Jaxb2Marshaller.java.html#L627)
6) [example 6](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/spring-framework/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/spring-framework/spring-web/src/main/java/org/springframework/web/util/HierarchicalUriComponents.java.html#L567)
7) [example 7](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/infinispan/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/infinispan/query/src/main/java/org/infinispan/query/backend/KeyTransformationHandler.java.html#L161)
8) [example 8](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/hibernate-orm-with-excludes/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/hibernate-orm-with-excludes/hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/Enhancer.java.html#L970)
9) [example 9](https://mezk.github.io/i1525-BooleanExpressionComplexity/diff/hibernate-orm-with-excludes/xref/home/aselkin/IDEA_Projects/contribution/checkstyle-tester/repositories/hibernate-orm-with-excludes/hibernate-entitymanager/src/main/java/org/hibernate/ejb/metamodel/AbstractManagedType.java.html#L240).

However, most of the examples show that the author of the code cannot find an appropriate name for the method to inline the logic. Is it a good reason for a new property? I don't think so.

**Comment by romani** [comment]
I think we need to detect pattern of expression and allow to skip validation if whole expression is same pattern. I do not know how to make a logic on this, as it very subjective and not very deterministic.
Does anyone have ideas on how to do this ? AI ? :) Or some complicated string diff on subexpression?

**Comment by romani** [comment]
@OldIMP, ∆∆

**Comment by nick-mancuso** [comment]
> Does anyone have ideas on how to do this? 

Yes, it probably isn't too bad, really. In cases like this, we would just have a chain of ancestor tokens that are all the same. So, we could have some property that allows same operator parents to only count as 1.

**Comment by ZaheerAhmadDev** [comment]
I am on it.
